File: src/demos/ros/demo_ROS_vehicle.cpp
// =============================================================================
// PROJECT CHRONO - http://projectchrono.org
//
// Copyright (c) 2026 projectchrono.org
// All rights reserved.
//
// Use of this source code is governed by a BSD-style license that can be found
// in the LICENSE file at the top level of the distribution and at
// http://projectchrono.org/license-chrono.txt.
//
// =============================================================================
// Authors: Aaron Young, Patrick Chen
// =============================================================================
//
// Demo: integrating ROS with Chrono::Vehicle. A JSON-specified wheeled vehicle
// (Audi) is driven from ROS (chrono_ros_interfaces/msg/DriverInputs) and its
// chassis state is published. Runs headless (no run-time window).
//
//   ros2 topic echo /chrono_ros_node/output/vehicle/state/pose
//   ros2 topic pub  /chrono_ros_node/input/driver_inputs ...
//
// =============================================================================

#include "chrono/core/ChTypes.h"

#include "chrono/physics/ChSystemNSC.h"
#include "chrono/physics/ChBody.h"

#include "chrono_ros/ChROSManager.h"
#include "chrono_ros/handlers/ChROSClockHandler.h"
#include "chrono_ros/handlers/ChROSBodyHandler.h"
#include "chrono_ros/handlers/vehicle/ChROSDriverInputsHandler.h"

#include "chrono_vehicle/ChConfigVehicle.h"
#include "chrono_vehicle/ChVehicleDataPath.h"
#include "chrono_vehicle/wheeled_vehicle/vehicle/WheeledVehicle.h"
#include "chrono_vehicle/terrain/RigidTerrain.h"
#include "chrono_vehicle/utils/ChVehicleUtilsJSON.h"
#include "chrono_vehicle/ChDriver.h"

using namespace chrono;
using namespace chrono::ros;
using namespace chrono::vehicle;

// JSON specification for an Audi wheeled vehicle.
static const std::string vehicle_json = "audi/json/audi_Vehicle.json";
static const std::string tire_json = "audi/json/audi_Pac02Tire.json";
static const std::string engine_json = "audi/json/audi_EngineSimpleMap.json";
static const std::string transmission_json = "audi/json/audi_AutomaticTransmissionSimpleMap.json";
static const std::string rigidterrain_json = "terrain/RigidPlane.json";

ChVector3d initLoc(0, 0, 0.5);
double initYaw = 20 * CH_DEG_TO_RAD;
double step_size = 2e-3;

int main(int argc, char* argv[]) {
    std::cout << "Copyright (c) 2026 projectchrono.org\nChrono version: " << CHRONO_VERSION << std::endl << std::endl;

    // Wheeled vehicle from JSON.
    WheeledVehicle vehicle(GetVehicleDataFile(vehicle_json), ChContactMethod::SMC);
    vehicle.Initialize(ChCoordsys<>(initLoc, QuatFromAngleZ(initYaw)));
    vehicle.GetChassis()->SetFixed(false);
    vehicle.SetChassisVisualizationType(VisualizationType::MESH);
    vehicle.SetChassisRearVisualizationType(VisualizationType::PRIMITIVES);
    vehicle.SetSuspensionVisualizationType(VisualizationType::PRIMITIVES);
    vehicle.SetSteeringVisualizationType(VisualizationType::PRIMITIVES);
    vehicle.SetWheelVisualizationType(VisualizationType::MESH);

    auto engine = ReadEngineJSON(GetVehicleDataFile(engine_json));
    auto transmission = ReadTransmissionJSON(GetVehicleDataFile(transmission_json));
    auto powertrain = chrono_types::make_shared<ChPowertrainAssembly>(engine, transmission);
    vehicle.InitializePowertrain(powertrain);

    for (auto& axle : vehicle.GetAxles()) {
        for (auto& wheel : axle->GetWheels()) {
            auto tire = ReadTireJSON(GetVehicleDataFile(tire_json));
            vehicle.InitializeTire(tire, wheel, VisualizationType::MESH);
        }
    }

    RigidTerrain terrain(vehicle.GetSystem(), GetVehicleDataFile(rigidterrain_json));
    terrain.Initialize();

    auto driver = std::make_shared<ChDriver>(vehicle);

    // ------------ ROS: clock + DriverInputs subscriber + chassis state.
    auto ros_manager = chrono_types::make_shared<ChROSManager>();
    ros_manager->RegisterHandler(chrono_types::make_shared<ChROSClockHandler>());
    ros_manager->RegisterHandler(
        chrono_types::make_shared<ChROSDriverInputsHandler>(25, driver, "~/input/driver_inputs"));
    ros_manager->RegisterHandler(
        chrono_types::make_shared<ChROSBodyHandler>(25, vehicle.GetChassisBody(), "~/output/vehicle/state"));
    ros_manager->Initialize();

    // ------------ Simulation loop (real-time paced by the vehicle).
    double t_end = 30;
    double time = 0;

    vehicle.EnableRealtime(true);
    while (time < t_end) {
        DriverInputs driver_inputs = driver->GetInputs();

        time = vehicle.GetSystem()->GetChTime();
        driver->Synchronize(time);
        vehicle.Synchronize(time, driver_inputs, terrain);
        terrain.Synchronize(time);

        driver->Advance(step_size);
        vehicle.Advance(step_size);
        terrain.Advance(step_size);

        if (!ros_manager->Update(time, step_size))
            break;
    }

    return 0;
}

File: src/demos/python/ros/demo_ROS_vehicle.py
# =============================================================================
# PROJECT CHRONO - http://projectchrono.org
#
# Copyright (c) 2026 projectchrono.org
# All rights reserved.
#
# Use of this source code is governed by a BSD-style license that can be found
# in the LICENSE file at the top level of the distribution and at
# http://projectchrono.org/license-chrono.txt.
#
# =============================================================================
# Authors: Aaron Young, Patrick Chen
# =============================================================================
#
# Chrono::Vehicle + ROS, in Python. An HMMWV is driven from ROS
# (chrono_ros_interfaces/msg/DriverInputs) and its chassis state is published.
# The ChROSDriverInputsHandler / ChROSBodyHandler are the same C++ built-ins used
# by the C++ demo, here taking a pychrono.vehicle ChDriver / ChBody.
#
#   ros2 topic echo /chrono_ros_node/output/vehicle/state/pose
#   ros2 topic pub  /chrono_ros_node/input/driver_inputs \
#       chrono_ros_interfaces/msg/DriverInputs "{throttle: 0.6}"
#
# =============================================================================

import pychrono as ch
import pychrono.vehicle as veh
import pychrono.ros as chros


def main():
    # Create the HMMWV vehicle, set parameters, and initialize.
    hmmwv = veh.HMMWV_Full()
    hmmwv.SetContactMethod(ch.ChContactMethod_NSC)
    hmmwv.SetChassisCollisionType(veh.CollisionType_NONE)
    hmmwv.SetChassisFixed(False)
    hmmwv.SetInitPosition(ch.ChCoordsysd(ch.ChVector3d(0, 0, 1.6), ch.ChQuaterniond(1, 0, 0, 0)))
    hmmwv.SetEngineType(veh.EngineModelType_SHAFTS)
    hmmwv.SetTransmissionType(veh.TransmissionModelType_AUTOMATIC_SHAFTS)
    hmmwv.SetDriveType(veh.DrivelineTypeWV_AWD)
    hmmwv.SetSteeringType(veh.SteeringTypeWV_PITMAN_ARM)
    hmmwv.SetTireType(veh.TireModelType_TMEASY)
    hmmwv.SetTireStepSize(1e-3)
    hmmwv.Initialize()

    # Create the terrain.
    terrain = veh.RigidTerrain(hmmwv.GetSystem())
    patch_mat = ch.ChContactMaterialNSC()
    patch_mat.SetFriction(0.9)
    patch_mat.SetRestitution(0.01)
    terrain.AddPatch(patch_mat, ch.CSYSNORM, 100.0, 100.0)
    terrain.Initialize()

    # Create the basic driver (commanded from ROS).
    driver = veh.ChDriver(hmmwv.GetVehicle())
    driver.Initialize()

    # ROS: clock + DriverInputs subscriber + chassis state publisher.
    ros_manager = chros.ChROSManager()
    ros_manager.RegisterHandler(chros.ChROSClockHandler())
    ros_manager.RegisterHandler(chros.ChROSDriverInputsHandler(25, driver, "~/input/driver_inputs"))
    ros_manager.RegisterHandler(chros.ChROSBodyHandler(25, hmmwv.GetChassisBody(), "~/output/vehicle/state"))
    ros_manager.Initialize()

    # Simulation loop (real-time paced).
    time = 0
    time_step = 1e-3
    time_end = 30

    hmmwv.GetVehicle().EnableRealtime(True)
    while time < time_end:
        time = hmmwv.GetSystem().GetChTime()

        driver_inputs = driver.GetInputs()

        driver.Synchronize(time)
        terrain.Synchronize(time)
        hmmwv.Synchronize(time, driver_inputs, terrain)

        driver.Advance(time_step)
        terrain.Advance(time_step)
        hmmwv.Advance(time_step)

        if not ros_manager.Update(time, time_step):
            break


if __name__ == "__main__":
    main()
